### ISSUE TITLE
Missing parentheses in generated query

### ISSUE THREAD

**Issue by milgner** [line/kotlin-jdsl#512]
The documentation states that the `and` and `or` predicates generate parentheses around their arguments.

However, this doesn't seem to be the case. Here's a function to look up all events newer than a given bookmark (which might not exist in the database, yet):

```
jpql {
    val mainEvent = entity(StoredEvent::class)
    val bookmarkEvent = entity(StoredEvent::class, "bookmarkEvent")
    val bookmarkProcessedAt = bookmarkEvent.path(StoredEvent::processedAt)
    val mainEventProcessedAt = mainEvent.path(StoredEvent::processedAt)
    
    select(mainEvent)
        .from(
            entity(EventStream::class),
            join(EventStream::events).`as`(mainEvent),
            leftJoin(Bookmark::class).on(path(Bookmark::name).eq(name)),
            leftJoin(Bookmark::lastEvent).`as`(bookmarkEvent)
        )
        .where(
            and(
                path(EventStream::name).eq(name),
                or(
                    path(Bookmark::lastEvent).isNull(),
                    and(
                        mainEventProcessedAt.greaterThanOrEqualTo(bookmarkProcessedAt),
                        mainEvent.path(StoredEvent::id).notEqual(bookmarkEvent.path(StoredEvent::id))
                    )
                )
            )
        )
        .orderBy(mainEventProcessedAt.asc())
}
```

The rendered query is

```
SELECT StoredEvent
FROM EventStream AS EventStream
         INNER JOIN EventStream.events AS StoredEvent
         LEFT JOIN Bookmark AS Bookmark ON Bookmark.name = :param1
         LEFT JOIN Bookmark.lastEvent AS bookmarkEvent
WHERE (EventStream.name = :param2) AND (Bookmark.lastEvent IS NULL)
   OR (StoredEvent.processedAt >= bookmarkEvent.processedAt) AND (StoredEvent.id <> bookmarkEvent.id)
ORDER BY StoredEvent.processedAt ASC
```

Instead I would have expected the parenthesis as follows:

```
SELECT StoredEvent FROM EventStream AS EventStream 
                    INNER JOIN EventStream.events AS StoredEvent
                    LEFT JOIN Bookmark AS Bookmark ON Bookmark.name = :param1
                    LEFT JOIN Bookmark.lastEvent AS bookmarkEvent
                WHERE (EventStream.name = :param2) AND (Bookmark.lastEvent IS NULL
   OR (StoredEvent.processedAt >= bookmarkEvent.processedAt AND StoredEvent.id <> bookmarkEvent.id))
```

**Comment by shouwn** [comment]
I have checked the cause. The cause is that the logic for removing duplicate parentheses only compares the start and end. 

I will fix the code and release 3.0.1 before next Monday. 

Sorry for the inconvenience.

**Comment by milgner** [comment]
No worries! Thank you for the quick reply! I wasn't sure whether I might have made a mistake or not.
For now I have switched the implementation to a HQL string and can easily change it back to jpql.

**Comment by shouwn** [comment]
I released 3.0.1 that solved the issue. Please check it. 🙏

**Comment by cj848** [comment]
I also experienced the same issue, so I changed the query to or -> in, but after changing it to 3.0.1, it was processed normally. This issue has been completely resolved.

my case is below.

3.0.0
```sql
where
        u1_0.id in(select
            max(u3_0.id) 
        from
            user_terms_agreement u3_0 
        join
            terms t2_0 
                on t2_0.id=u3_0.terms_id 
        where
            (u3_0.user_id=2) 
            and (t2_0.category='TERMS_OF_SERVICE') 
            or (t2_0.category='SNOW_POLICY') 
        group by
            t2_0.id)

```

3.0.1
```sql
where
        u1_0.id in(select
            max(u3_0.id) 
        from
            user_terms_agreement u3_0 
        join
            terms t2_0 
                on t2_0.id=u3_0.terms_id 
        where
            (u3_0.user_id=3) 
            and ((t2_0.category='TERMS_OF_SERVICE') 
            or (t2_0.category='SNOW_POLICY')) 
        group by
            t2_0.id)
```
